### PR TITLE
Remove ast-ty-to-ty cache

### PR DESCRIPTION
**Pull request by nikomatsakis**
As discussed on IRC, this basically just removes the cache, and rewrites rustdoc and save-analysis so call into the astconv code. It *might* make sense for this to be a more fine-grained query, but that would (at least) require us to be using `HirId` and not `NodeId`.

(Perhaps I should open a FIXME?)

I didn't measure perf impact (yet?). I did observe that the cache seems to hit *rarely* -- and only in between items (I experimented with a cache "per def-id", but that had zero hits). In other words, every single hit on the cache is a dependency bug, since it is "shuttling" information between items without dependency edges.

r? @eddyb

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
FYI: The perf impact will be measured for this specific PR as long as it isn't rolled up, so I wouldn't be too concerned with measuring separately.

File: src/librustc_save_analysis/dump_visitor.rs [hunk 1]
```diff
@@ -115,15 +117,18 @@ impl<'l, 'tcx: 'l, 'll, D: Dump + 'll> DumpVisitor<'l, 'tcx, 'll, D> {
         where F: FnOnce(&mut DumpVisitor<'l, 'tcx, 'll, D>)
     {
         let item_def_id = self.tcx.hir.local_def_id(item_id);
+        let old_item_cx = mem::replace(&mut self.save_ctxt.item_cx,
+                                       ItemCtxt::new(self.tcx, item_def_id));
```
Comment (eddyb) [inline src/librustc_save_analysis/dump_visitor.rs]: Hah I didn't expect this. I'm not sure we should have this at all. That is, `ItemCtxt` is just a `(TyCtxt, DefId)` pair at this point IIRC, and the `item_def_id` seems to be the important thing to keep?

Comment (nikomatsakis) [inline src/librustc_save_analysis/dump_visitor.rs]: Yeah, I was debating about stripping it out too -- I was thinking I could just make a `pub fn` in `librustc_typeck` that takes a def-id and constructs a "fresh itemctxt". I think when I wrote this patch I was anticipating having a cache per `ItemCtxt` (but then I checked and found that said cache had zero hits).

File: src/librustdoc/clean/mod.rs [hunk 2]
```diff
@@ -53,51 +55,71 @@ use html::item_type::ItemType;
 pub mod inline;
 mod simplify;
 
+pub struct CleanContext<'a, 'tcx: 'a>  {
+    doc_cx: &'a DocContext<'a, 'tcx>,
+    item_cx: ItemCtxt<'a, 'tcx>,
```
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: This is... unnecessary? Same thing as before, we *really* don't need to inflate the diff like this.

File: src/librustc_save_analysis/dump_visitor.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc_save_analysis/dump_visitor.rs]: > I was thinking I could just make a `pub fn` in `librustc_typeck` that takes a def-id and constructs a "fresh itemctxt"

That's... what `ItemCtxt::new` is, isn't it?

**Comment by nikomatsakis** [comment]
@eddyb ok, take another look.

**Comment by TimNN** [comment]
@bors retry

* #41545

**Pull request merged**